src/player.h

```diff
@@ -1862,6 +1862,9 @@ class player : public Character
         /** Stamp of skills. @ref learned_recipes are valid only with this set of skills. */
         mutable decltype( _skills ) valid_autolearn_skills;
 
+        /** Amount of time the player has spent in each overmap tile. */
+        std::unordered_map<tripoint, time_duration> overmap_time;
```

Comment:
This hash-map is a time-bomb (pun intended): it's going to gobble memory never giving it back. Other drawbacks are bloated save-files and possible performance losses. I'd advise you against following that path. As an alternative you could use some kind of an anchor (one point), a "center of mass" of all recently visited overmap tiles, which slowly chases the player. The closer you are to that anchor, the higher the morale penalty. I guess this would work.

Reply:
Plus the more I think about it the less a perpetually stacking pile of negative morale makes sense. Birds, fish and nomadic human cultures tend to have cyclical migration structures returning to the same feeding/spawning grounds (or trade hubs for humans) every year. Processing this as a "decay" per overmap tile would be very inefficient, the "anchor" idea sounds great (and could potentially be used in the future for a Nemesis-style super zombie pursuer...)

Reply:
I disagree with the hash map concern, it's a nicely generic piece of data, and it's resolution leaves it producing a very small amount of data.
Consider a pathological use case of continuously walking in some direction.
Each entry is approximately 128 bits/16 bytes (three 32 bit ints for a tripoint, and one 32 bit int for a counter).
We create a new entry once every 24 turns (the diameter of an OMT). That's roughly 2 minutes.
After 5 years of game time, that's 1.3 million entries.
That's about 16MB of data assuming we never ever prune it.
If we segment by overmap, this becomes a non-issue.

Reply:
@kevingranade It does only count awake time on each OMT -- does that make it as nicely generic?

Reply:
> I disagree with the hash map concern, it's a nicely generic piece of data, and it's resolution leaves it producing a very small amount of data.

I'm certainly not opposed to hash maps in general (they're cool), but I'm still against using one here.

> Each entry is approximately 128 bits/16 bytes (three 32 bit ints for a tripoint, and one 32 bit int for a counter).

Let me elaborate your estimations a bit. There's at least 64 bits of data per entry additionally
(due to the implementation details): a pointer in a single-linked list and an iterator in the table referencing it. This gives us 24 bytes per entry (31.5MB of data) plus some additional memory for the table growth (say 30%, or 1.5MB). So it's more like 33MB of memory which can become roughly three times the size when serialized as strings (my last assumption is pretty much handwavy, but still).

My point is, we don't have to pay this price at all.